libnavigation-core/src/main/java/com/mapbox/navigation/MapboxNavigation.kt

```diff
@@ -0,0 +1,355 @@
+                val originTolerance =
+                    bearingElements?.getOrNull(0)?.split(",")?.getOrNull(1)?.toDouble()
+                bearingElements?.subList(index + 1, coordinates.size)?.map { element ->
+                    element.split(",").let { components ->
+                        if (components.size == 2) {
+                            Pair(components[0].toDouble(), components[1].toDouble())
+                        } else {
+                            null
+                        }
+                    }
+                }?.toMutableList()?.also { pairs ->
+                    pairs.add(
+                        0,
+                        Pair(
+                            location.bearing.toDouble(),
+                            originTolerance ?: DEFAULT_REROUTE_BEARING_TOLERANCE
+                        )
+                    )
+                    optionsBuilder.bearings(*pairs.toTypedArray())
+                }
+
+                // todo implement options.radiuses
+                // todo implement options.approaches
+                // todo implement options.waypointIndices
+                // todo implement options.waypointNames
+                // todo implement options.waypointTargets
+            }
+
+            val optionsRebuilt = optionsBuilder.build()
+            directionsSession.requestRoutes(optionsRebuilt)
+        }
+    }
+
+    private fun monitorNotificationActionButton(channel: ReceiveChannel<NotificationAction>) {
+        mainJobController.scope.monitorChannelWithException(channel) { notificationAction ->
+            when (notificationAction) {
+                NotificationAction.END_NAVIGATION -> tripSession.stop()
+            }
+        }
+    }
+
+    /**
+     * Provides parameters for Mapbox default modules, recursively if a module depends on other Mapbox modules.
+     */
+    private fun paramsProvider(type: MapboxNavigationModuleType): Array<Pair<Class<*>?, Any?>> {
+        return when (type) {
+            MapboxNavigationModuleType.HybridRouter -> arrayOf(
+                Router::class.java to NavigationModuleProvider.createModule(
+                    MapboxNavigationModuleType.OnboardRouter,
+                    ::paramsProvider
+                ),
+                Router::class.java to NavigationModuleProvider.createModule(
+                    MapboxNavigationModuleType.OffboardRouter,
+                    ::paramsProvider
+                )
+            )
+            MapboxNavigationModuleType.OffboardRouter -> arrayOf(
+                String::class.java to (accessToken
+                    ?: throw RuntimeException("You need to provide an access in order to use the default OffboardRouter.")),
+                Context::class.java to context
+            )
+            MapboxNavigationModuleType.OnboardRouter -> arrayOf()
+            MapboxNavigationModuleType.DirectionsSession -> throw NotImplementedError() // going to be removed when next base version
```

Comment:
`TODO` Same for some other modules below 👇

Reply:
https://github.com/mapbox/mapbox-base-android/pull/23